x/gamm/pool-models/balancer/amm_test.go

```diff
@@ -324,3 +318,65 @@ func TestCalcSingleAssetInAndOut_InverseRelationship(t *testing.T) {
 		}
 	}
 }
+
+func TestCalcJoinPoolShares(t *testing.T) {
+	testCases := []struct {
+		name         string
+		swapFee      sdk.Dec
+		poolAssets   []balancer.PoolAsset
+		tokensIn     sdk.Coins
+		expectErr    bool
+		expectShares sdk.Int
+		expectLiq    sdk.Coins
+	}{
+		{
+			name:    "equal weights with zero swap fee",
+			swapFee: sdk.MustNewDecFromStr("0"),
+			poolAssets: []balancer.PoolAsset{
+				{
+					Token:  sdk.NewInt64Coin("uosmo", 1_000_000_000_000),
+					Weight: sdk.NewInt(100),
+				},
+				{
+					Token:  sdk.NewInt64Coin("uatom", 1_000_000_000_000),
+					Weight: sdk.NewInt(100),
+				},
+			},
+			tokensIn:     sdk.NewCoins(sdk.NewInt64Coin("uosmo", 50_000)),
+			expectErr:    false,
+			expectShares: sdk.NewInt(2499999968800),
+			expectLiq:    sdk.NewCoins(sdk.NewInt64Coin("uosmo", 50000)),
+		},
+		// {
+		// 	name: "equal weights with 0.001 swap fee",
+		// 	swapFee: sdk.MustNewDecFromStr("0.001"),
+		// },
+		// {
+		// 	name: "equal weights with 0.1 swap fee",
+		// 	swapFee: sdk.MustNewDecFromStr("0.1"),
+		// },
+		// {
+		// 	name: "equal weights with 0.99 swap fee",
+		// 	swapFee: sdk.MustNewDecFromStr("0.99"),
+		// },
+	}
+
+	for _, tc := range testCases {
+		tc := tc
```

Comment:
Is this to ensure that we are operating on the desired copy when the for loop advances to the next tc?

Reply:
Correct, since we run the test in a closure.